agent/auto-config/auto_config.go

```diff
@@ -0,0 +1,480 @@
+	if err != nil {
+		return nil, err
+	}
+
+	if !ready {
+		if err := ac.getInitialConfiguration(ctx); err != nil {
+			return nil, err
+		}
+	}
+
+	// re-read the configuration now that we have our initial auto-config
+	config, err := ac.ReadConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	ac.config = config
+	return ac.config, nil
+}
+
+// introToken is responsible for determining the correct intro token to use
+// when making the initial Cluster.AutoConfig RPC request.
+func (ac *AutoConfig) introToken() (string, error) {
+	conf := ac.config.AutoConfig
+	// without an intro token or intro token file we cannot do anything
+	if conf.IntroToken == "" && conf.IntroTokenFile == "" {
+		return "", fmt.Errorf("neither intro_token or intro_token_file settings are not configured")
+	}
+
+	token := conf.IntroToken
+	if token == "" {
+		// load the intro token from the file
+		content, err := ioutil.ReadFile(conf.IntroTokenFile)
+		if err != nil {
+			return "", fmt.Errorf("Failed to read intro token from file: %w", err)
+		}
+
+		token = string(content)
+
+		if token == "" {
+			return "", fmt.Errorf("intro_token_file did not contain any token")
+		}
+	}
+
+	return token, nil
+}
+
+// autoConfigHosts is responsible for taking the list of server addresses and
+// resolving any go-discover provider invocations. It will then return a list
+// of hosts. These might be hostnames and is expected that DNS resolution may
+// be performed after this function runs. Additionally these may contain ports
+// so SplitHostPort could also be necessary.
+func (ac *AutoConfig) autoConfigHosts() ([]string, error) {
+	servers := ac.config.AutoConfig.ServerAddresses
+
+	providers := make(map[string]discover.Provider)
+	for k, v := range discover.Providers {
+		providers[k] = v
+	}
+	providers["k8s"] = &discoverk8s.Provider{}
+
+	disco, err := discover.New(
+		discover.WithUserAgent(lib.UserAgent()),
+		discover.WithProviders(providers),
+	)
+
+	if err != nil {
+		return nil, fmt.Errorf("Failed to create go-discover resolver: %w", err)
+	}
+
+	var addrs []string
+	for _, addr := range servers {
+		switch {
+		case strings.Contains(addr, "provider="):
+			resolved, err := disco.Addrs(addr, ac.logger.StandardLogger(&hclog.StandardLoggerOptions{InferLevels: true}))
+			if err != nil {
+				ac.logger.Error("failed to resolve go-discover auto-config servers", "configuration", addr, "err", err)
+				continue
+			}
+
+			addrs = append(addrs, resolved...)
+			ac.logger.Debug("discovered auto-config servers", "servers", resolved)
+		default:
+			addrs = append(addrs, addr)
+		}
+	}
+
+	if len(addrs) == 0 {
+		return nil, fmt.Errorf("no auto-config server addresses available for use")
+	}
+
+	return addrs, nil
+}
+
+// resolveHost will take a single host string and convert it to a list of TCPAddrs
+// This will process any port in the input as well as looking up the hostname using
+// normal DNS resolution.
+func (ac *AutoConfig) resolveHost(hostPort string) []net.TCPAddr {
```

Comment:
Did you take that from auto_encrypt? That part was actually a little hard to make it work similar to memberlist.

Reply:
Not really, I didn't bother trying to make it look similar to memberlist with the whole make a manual TCP request followed by a UDP one. These have to be server addresses and at least 15-20 AAAA and even more A records can be returned for a plain DNS query response. That number should be way more than the number of servers in the cluster so truncation isn't an issue. Even then its likely EDNS will be in use in which case we have even more space at our disposal.

So this function just takes a string which is a host with an optional `:<port>` and resolves the host to IPs and makes a list of all the `TCPAddrs` using the specified port or the default server port otherwise.

Its using the default system resolver for the IP lookup.

Reply:
Sounds good to me!